File: backend/worktree.py
"""Git worktree helpers — create/remove per-task worktrees."""

from __future__ import annotations

import asyncio
import logging
import os
from pathlib import Path
from typing import Optional, Tuple

logger = logging.getLogger(__name__)


async def _run_git(args: list[str], cwd: str) -> Tuple[int, str, str]:
    """Run a git command and return (returncode, stdout, stderr)."""
    proc = await asyncio.create_subprocess_exec(
        "git", *args,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    stdout, stderr = await proc.communicate()
    return proc.returncode, stdout.decode().strip(), stderr.decode().strip()


def is_git_repo(path: str) -> bool:
    """Check if path is inside a git repository."""
    return os.path.isdir(os.path.join(path, ".git"))


async def get_default_branch(repo_path: str) -> str:
    """Detect the default branch (main or master)."""
    rc, out, _ = await _run_git(
        ["symbolic-ref", "refs/remotes/origin/HEAD", "--short"], repo_path
    )
    if rc == 0 and out:
        return out.split("/")[-1]
    # Fallback: check master first, then main
    rc, _, _ = await _run_git(["rev-parse", "--verify", "master"], repo_path)
    if rc == 0:
        return "master"
    return "main"


def worktree_base_dir(repo_path: str) -> Path:
    return Path(repo_path) / ".worktrees"


def worktree_path_for_task(repo_path: str, task_id: str) -> Path:
    return worktree_base_dir(repo_path) / task_id


def branch_name_for_task(task_id: str) -> str:
    return f"task/{task_id}"


async def create_worktree(
    repo_path: str, task_id: str
) -> Tuple[bool, str, Optional[str]]:
    """Create a git worktree for a task.

    Returns (success, worktree_path_or_error, branch_name).
    """
    if not is_git_repo(repo_path):
        return False, f"Not a git repository: {repo_path}", None

    wt_path = worktree_path_for_task(repo_path, task_id)
    branch = branch_name_for_task(task_id)

    # Already exists (e.g. task restart)
    if wt_path.exists():
        return True, str(wt_path), branch

    # Ensure .worktrees/ is excluded from git tracking
    _ensure_excluded(repo_path)

    default_branch = await get_default_branch(repo_path)

    # Create worktree with new branch from default branch
    rc, out, err = await _run_git(
        ["worktree", "add", "-b", branch, str(wt_path), default_branch],
        repo_path,
    )
    if rc != 0:
        # Branch might already exist (task restarted after worktree removed)
        if "already exists" in err:
            rc2, _, err2 = await _run_git(
                ["worktree", "add", str(wt_path), branch], repo_path
            )
            if rc2 != 0:
                return False, f"Failed to create worktree: {err2}", None
        else:
            return False, f"Failed to create worktree: {err}", None

    logger.info("Created worktree for task %s: %s (branch: %s)", task_id, wt_path, branch)
    return True, str(wt_path), branch


async def remove_worktree(repo_path: str, task_id: str) -> bool:
    """Remove a worktree (keeps the branch)."""
    wt_path = worktree_path_for_task(repo_path, task_id)
    if not wt_path.exists():
        return True

    rc, _, err = await _run_git(
        ["worktree", "remove", str(wt_path), "--force"], repo_path
    )
    if rc != 0:
        logger.warning("Failed to remove worktree %s: %s", wt_path, err)
        return False

    logger.info("Removed worktree for task %s", task_id)
    return True


def _ensure_excluded(repo_path: str) -> None:
    """Add .worktrees/ to .git/info/exclude if not already there."""
    exclude = Path(repo_path) / ".git" / "info" / "exclude"
    if not exclude.exists():
        return
    try:
        content = exclude.read_text()
        if ".worktrees" not in content:
            with open(exclude, "a") as f:
                f.write("\n.worktrees/\n")
    except OSError:
        pass

File: backend/src/worktree.ts
/**
 * Git worktree helpers — create/remove per-task worktrees.
 * Port of backend/worktree.py to TypeScript.
 */

import { execFile } from "child_process";
import fs from "fs";
import fsp from "fs/promises";
import path from "path";

function runGit(args: string[], cwd: string): Promise<{ code: number; stdout: string; stderr: string }> {
  return new Promise((resolve) => {
    execFile("git", args, { cwd }, (error, stdout, stderr) => {
      resolve({
        code: error ? 1 : 0,
        stdout: (stdout ?? "").trim(),
        stderr: (stderr ?? "").trim(),
      });
    });
  });
}

export function isGitRepo(repoPath: string): boolean {
  return fs.existsSync(path.join(repoPath, ".git"));
}

export async function getDefaultBranch(repoPath: string): Promise<string> {
  const { code, stdout } = await runGit(
    ["symbolic-ref", "refs/remotes/origin/HEAD", "--short"],
    repoPath,
  );
  if (code === 0 && stdout) return stdout.split("/").pop()!;

  // Fallback: check master first, then main
  const { code: rc } = await runGit(["rev-parse", "--verify", "master"], repoPath);
  return rc === 0 ? "master" : "main";
}

export function worktreeBaseDir(repoPath: string): string {
  return path.join(repoPath, ".worktrees");
}

export function worktreePathForTask(repoPath: string, taskId: string): string {
  return path.join(worktreeBaseDir(repoPath), taskId);
}

export function branchNameForTask(taskId: string): string {
  return `task/${taskId}`;
}

/**
 * Create a git worktree for a task.
 * Returns [success, worktree_path_or_error, branch_name].
 */
export async function createWorktree(
  repoPath: string,
  taskId: string,
): Promise<[boolean, string, string | null]> {
  if (!isGitRepo(repoPath)) {
    return [false, `Not a git repository: ${repoPath}`, null];
  }

  const wtPath = worktreePathForTask(repoPath, taskId);
  const branch = branchNameForTask(taskId);

  // Already exists (e.g. task restart)
  if (fs.existsSync(wtPath)) {
    return [true, wtPath, branch];
  }

  // Ensure .worktrees/ is excluded from git tracking
  ensureExcluded(repoPath);

  const defaultBranch = await getDefaultBranch(repoPath);

  // Create worktree with new branch from default branch
  const { code, stderr } = await runGit(
    ["worktree", "add", "-b", branch, wtPath, defaultBranch],
    repoPath,
  );

  if (code !== 0) {
    // Branch might already exist (task restarted after worktree removed)
    if (stderr.includes("already exists")) {
      const { code: rc2, stderr: err2 } = await runGit(
        ["worktree", "add", wtPath, branch],
        repoPath,
      );
      if (rc2 !== 0) return [false, `Failed to create worktree: ${err2}`, null];
    } else {
      return [false, `Failed to create worktree: ${stderr}`, null];
    }
  }

  console.log(`[worktree] Created worktree for task ${taskId}: ${wtPath} (branch: ${branch})`);
  return [true, wtPath, branch];
}

export async function removeWorktree(repoPath: string, taskId: string): Promise<boolean> {
  const wtPath = worktreePathForTask(repoPath, taskId);
  if (!fs.existsSync(wtPath)) return true;

  const { code, stderr } = await runGit(
    ["worktree", "remove", wtPath, "--force"],
    repoPath,
  );
  if (code !== 0) {
    console.warn(`[worktree] Failed to remove worktree ${wtPath}: ${stderr}`);
    return false;
  }

  console.log(`[worktree] Removed worktree for task ${taskId}`);
  return true;
}

function ensureExcluded(repoPath: string): void {
  const exclude = path.join(repoPath, ".git", "info", "exclude");
  if (!fs.existsSync(exclude)) return;
  try {
    const content = fs.readFileSync(exclude, "utf-8");
    if (!content.includes(".worktrees")) {
      fs.appendFileSync(exclude, "\n.worktrees/\n");
    }
  } catch {
    // ignore
  }
}
